### PR TITLE
Add keycloak_user_federation module

### LINKED ISSUE #3027: Add support for keycloak User Federation
### Summary

I would love to have support for configuring User Federation and it's (ldap) mappers in keycloak. 

### Issue Type

Feature Idea

### Component Name

keycloak_user_federation

### Additional Information

This would allow for a playbook that install's and configures a keycloak server with ldap integration. 

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by laurpaum**
##### SUMMARY

Add a new module to manage Keycloak user federations (LDAP and Kerberos)
Fixes #3027

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

keycloak_user_federation

### CODE REVIEW

File: plugins/modules/identity/keycloak/keycloak_user_federation.py [hunk 1]
```diff
@@ -0,0 +1,977 @@
...
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+DOCUMENTATION = '''
+---
+module: keycloak_user_federation
+
+short_description: Allows administration of Keycloak user federations via Keycloak API
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_user_federation.py:13]:
```suggestion
short_description: Allows administration of Keycloak user federations via Keycloak API

version_added: 3.7.0
```

File: plugins/modules/identity/keycloak/keycloak_user_federation.py [hunk 2]
```diff
@@ -0,0 +1,977 @@
...
+    realm:
+        description:
+            - The Keycloak realm under which this user federation resides.
+        default: 'master'
+        type: str
+
+    id:
+        description:
+            - The unique ID for this user federation. If left empty, the user federation will be searched
+              by its `name`.
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_user_federation.py:48]:
```suggestion
              by its I(name).
```

File: plugins/modules/identity/keycloak/keycloak_user_federation.py [hunk 3]
```diff
@@ -0,0 +1,977 @@
...
+        type: str
+
+    name:
+        description:
+            - Display name of provider when linked in admin console.
+        type: str
+
+    provider_id:
+        description:
+            - Provider for this user federation (supported values are C(ldap) or C(kerberos)).
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_user_federation.py:58]:
```suggestion
            - Provider for this user federation.
```
The allowed values are listed next to this option, so I think it's ok to not repeat them. (Also this makes it easier to not forget updating this list in the future, if it's ever extended :) ).

File: plugins/modules/identity/keycloak/keycloak_user_federation.py [hunk 4]
```diff
@@ -0,0 +1,977 @@
...
+            - Dict specifying the configuration options for the provider; the contents differ depending on
+              the value of I(provider_id). Examples are given below for C(ldap) and C(kerberos). It is easiest
+              to obtain valid config values by dumping an already-existing user federation configuration
+              through check-mode in the I(existing) field.
+        type: dict
+        suboptions:
+            enabled:
+                description:
+                    - Enable/disable this user federation.
+                default: True
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_user_federation.py:92]:
```suggestion
                default: true
```
Always use `true` and `false` for YAML boolean.s

File: plugins/modules/identity/keycloak/keycloak_user_federation.py [hunk 5]
```diff
@@ -0,0 +1,977 @@
...
+
+            priority:
+                description:
+                    - Priority of provider when doing a user lookup. Lowest first.
+                default: 0
+                type: int
+
+            importEnabled:
+                description:
+                    - If C(True), LDAP users will be imported into Keycloak DB and synced by the configured
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_user_federation.py:103]:
```suggestion
                    - If C(true), LDAP users will be imported into Keycloak DB and synced by the configured
```
(Also in corresponding descriptions.)

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution! Here are a few first comments.

File: plugins/modules/identity/keycloak/keycloak_user_federation.py [hunk 6]
```diff
@@ -0,0 +1,979 @@
...
+      auth_password: password
+      realm: my-realm
+      name: my-federation
+      state: absent
+
+'''
+
+RETURN = '''
+msg:
+    description: Message as to what action was taken
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_user_federation.py:548]:
```suggestion
    description: Message as to what action was taken.
```

File: plugins/modules/identity/keycloak/keycloak_user_federation.py [hunk 7]
```diff
@@ -0,0 +1,979 @@
...
+    returned: always
+    type: str
+    sample: "No changes required to user federation 164bb483-c613-482e-80fe-7f1431308799."
+
+proposed:
+    description: Representation of proposed changes to user federation
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_user_federation.py:554]:
```suggestion
    description: Representation of proposed changes to user federation.
```

File: plugins/modules/identity/keycloak/keycloak_user_federation.py [hunk 8]
```diff
@@ -0,0 +1,979 @@
...
+            "validatePasswordPolicy": "false",
+            "vendor": "other"
+        },
+        "name": "ldap",
+        "providerId": "ldap",
+        "providerType": "org.keycloak.storage.UserStorageProvider"
+    }
+
+existing:
+    description: Representation of existing user federation
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_user_federation.py:592]:
```suggestion
    description: Representation of existing user federation.
```

File: plugins/modules/identity/keycloak/keycloak_user_federation.py [hunk 9]
```diff
@@ -0,0 +1,979 @@
...
+            }
+        ],
+        "name": "myfed",
+        "parentId": "myrealm",
+        "providerId": "ldap",
+        "providerType": "org.keycloak.storage.UserStorageProvider"
+    }
+
+end_state:
+    description: Representation of user federation after module execution
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_user_federation.py:650]:
```suggestion
    description: Representation of user federation after module execution.
```

File: plugins/modules/identity/keycloak/keycloak_user_federation.py [hunk 10]
```diff
@@ -0,0 +1,979 @@
...
+                type: bool
+
+            useTruststoreSpi:
+                description:
+                    - Specifies whether LDAP connection will use the truststore SPI with the truststore
+                      configured in standalone.xml/domain.xml. C(Always) means that it will always use it.
+                      C(Never) means that it will not use it. C(Only for ldaps) means that it will use if
+                      your connection URL use ldaps. Note even if standalone.xml/domain.xml is not
+                      configured, the default Java cacerts or certificate specified by
+                      'javax.net.ssl.trustStore' property will be used.
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_user_federation.py:248]:
```suggestion
                      C(javax.net.ssl.trustStore) property will be used.
```

File: plugins/modules/identity/keycloak/keycloak_user_federation.py [hunk 11]
```diff
@@ -0,0 +1,979 @@
...
+            searchScope:
+                description:
+                    - For one level, the search applies only for users in the DNs specified by User DNs.
+                      For subtree, the search applies to the whole subtree. See LDAP documentation for
+                      more details
+                default: '1'
+                type: str
+                choices:
+                    - 1
+                    - 2
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_user_federation.py:190]:
```suggestion
                    - '1'
                    - '2'
```

**Comment by felixfontein** [comment]
Just a few more docs comments. Code looks good as far as I can judge it.

**Comment by laurpaum** [comment]
@felixfontein thanks for your feedback

**Review by felixfontein — APPROVED** [review]
Looks good to me. If nobody complains, I'll merge this by the beginning of next week.

**Comment by felixfontein** [comment]
CC @Gaetan2907 @kris2kris @pmdumuid

File: plugins/modules/identity/keycloak/keycloak_user_federation.py [hunk 12]
```diff
@@ -0,0 +1,979 @@
...
+                if len(found) == 1:
+                    old_mapper = found[0]
+                else:
+                    old_mapper = dict()
+            new_mapper = old_mapper.copy()
+            new_mapper.update(change)
+            if new_mapper != old_mapper:
+                if changeset.get('mappers') is None:
+                    changeset['mappers'] = list()
+                changeset['mappers'].append(new_mapper)
```
Comment (pmdumuid) [inline plugins/modules/identity/keycloak/keycloak_user_federation.py:872]: Do we handle removing of mappers?  Could we add that to the tests below? (e.g. create a user_federation, with a mapper then delete one of them)

**Review by pmdumuid — CHANGES_REQUESTED** [review]
Can you please add a test to show adding and removing of mappers.

File: plugins/modules/identity/keycloak/keycloak_user_federation.py [hunk 13]
```diff
@@ -0,0 +1,979 @@
...
+    # convert module parameters to client representation parameters (if they belong in there)
+    comp_params = [x for x in module.params
+                   if x not in list(keycloak_argument_spec().keys()) + ['state', 'realm', 'mappers'] and
+                   module.params.get(x) is not None]
+
+    # does the user federation already exist?
+    if cid is None:
+        found = kc.get_components('type=org.keycloak.storage.UserStorageProvider&parent={id}&name={name}'.format(id=realm, name=name), realm)
+        if len(found) > 1:
+            module.fail_json(msg='No ID given and found multiple user federations with name `{name}`. Cannot continue.'.format(name=name))
```
[the remark this answers is not in the document]
Comment (laurpaum) [reply plugins/modules/identity/keycloak/keycloak_user_federation.py:820]: Yes, `fail_json` calls `sys.exit`, no need to add an explicit `return`

File: plugins/modules/identity/keycloak/keycloak_user_federation.py [hunk 14]
```diff
@@ -0,0 +1,979 @@
...
+            cid = before_comp['id']
+    else:
+        before_comp = kc.get_component(cid, realm)
+
+    if before_comp is None:
+        before_comp = dict()
+
+    # if user federation exists, get associated mappers
+    if cid is not None:
+        before_comp['mappers'] = sorted(kc.get_components(urlencode(dict(parent=cid)), realm), key=lambda x: x.get('name'))
```
[the remark this answers is not in the document]
Comment (laurpaum) [reply plugins/modules/identity/keycloak/keycloak_user_federation.py:832]: That's the way the Keycloak API works. There are no specific endpoints for user federations or LDAP mappers. These are both handled as "component" objects with a hierarchical relation.

File: plugins/modules/identity/keycloak/keycloak_user_federation.py [hunk 15]
```diff
@@ -0,0 +1,979 @@
...
+        old_value = before_comp[camel(param)] if camel(param) in before_comp else None
+        if param == 'mappers':
+            new_param_value = [dict((k, v) for k, v in x.items() if x[k] is not None) for x in new_param_value]
+        if new_param_value != old_value:
+            changeset[camel(param)] = new_param_value
+
+    # special handling of mappers list to allow change detection
+    if module.params.get('mappers') is not None:
+        if module.params['provider_id'] == 'kerberos':
+            module.fail_json(msg='Cannot configure mappers for Kerberos federations.')
```
[the remark this answers is not in the document]
Comment (laurpaum) [reply plugins/modules/identity/keycloak/keycloak_user_federation.py:848]: The module supports both LDAP and Kerberos user federations. Kerberos federation only handles user authentication, you cannot add mappers to a Kerberos federation.

File: plugins/modules/identity/keycloak/keycloak_user_federation.py (hunk 12 above)
Comment (laurpaum) [reply plugins/modules/identity/keycloak/keycloak_user_federation.py:872]: For now, I only handled mappers creation, not removal. I can add the feature if there's a need for it.

**Comment by felixfontein** [comment]
@pmdumuid you're still requiring changes. Is the current state good enough for a first version to be merged and released?

File: plugins/modules/identity/keycloak/keycloak_user_federation.py (hunk 12 above)
Comment (pmdumuid) [reply plugins/modules/identity/keycloak/keycloak_user_federation.py:872]: Fair enough.. Happy for this to go as is.  I might follow up on this concept myself as I may have need for this feature myself.

**Comment by pmdumuid** [comment]
@felixfontein  - I'm happy with what I've seen.

**Comment by felixfontein** [comment]
Awesome, thanks! I'll merge this evening if nobody else complains.

**Comment by felixfontein** [comment]
@laurpaum thanks for contributing this module!
@pmdumuid thanks for reviewing!

**Pull request merged**